fix PR 7150 - reproducible builds

WalterBright opened the pull request:
Fixes for https://github.com/dlang/dmd/pull/7150

1. initial size of fixup array is too small
2. fixes to work with DMC++ front end

Depends on #7165

WalterBright committed:
fix PR 7150 - reproducible builds

- initial size of fixup array is too small
- fixes to work with DMC++ front end

ZombineDev commented on src/ddmd/backend/cgen.c:
```diff
@@ -812,6 +813,13 @@ size_t addtofixlist(symbol *s,targ_size_t offset,int seg,targ_size_t val,int fla
+{
+    //printf("searchfixlist(%s)\n", s->Sident);
+}
+#endif
```
Looks like this can be removed.

WalterBright replied:
Left as a convenient debugging feature.

ZombineDev commented on src/ddmd/backend/code.d:
```diff
@@ -217,6 +217,7 @@ extern __gshared int refparam;
 targ_size_t cod3_spoff();
 uint calccodsize(code *c);
 targ_size_t cod3_bpoffset(Symbol *s);
+void searchfixlist(Symbol *s) { }
```
This too.

WalterBright replied:
#7150 explains why that is there.

ZombineDev replied:
Thanks, I missed Martin's comment there.

ibuclaw commented:
Looking at travisci, it seems like something is still not deterministic.

stefan-koch-sociomantic commented on src/ddmd/backend/cgen.c:
```diff
@@ -743,7 +743,8 @@ struct FixupArray
         {
-            cap = cap ? 2 * cap : 16;
+            // 0x800 determined experimentally to minimize reallocations
+            cap = cap ? 2 * cap : 0x800;
```
2048 ?

MartinNowak committed:
fix uninitialized memory access

- types `ty0` and `ty1` are only lazily initialized with the
  first expression accessing a sliced field
- fixed by checking the workaround condition after all gathering

MartinNowak commented:
Why didn't you wait on me or at least check the test results?

MartinNowak committed:
show nm diff in case binaries aren't identical

- fixes failure on OSX which doesn't have sha1sum

andralex commented on src/ddmd/backend/cgen.c:
```diff
@@ -709,7 +709,10 @@ struct FixupArray
-            cap = cap ? 2 * cap : 16;
+            // 0x800 determined experimentally to minimize reallocations
+            cap = cap
+                ? (3 * cap) / 2 // use 'Tau' of 1.5
```
I assume cap never decreases, so it's either 0 or something no less than 0x800. Otherwise the code has a bug if cap = 1.

WalterBright replied:
`cap` starts at 0, and is set to 0x800 upon first use. It is never `0<cap<0x800`.

andralex commented:
please take a look at the test failures
